Add `type` parameter to `/system/df`

rvolosatovs opened the pull request:
**- What I did**
Add `type` parameter to `/system/df`

**- How I did it**
Internally allow more fine-grained control over choice of objects to compute disk usage from, add `types` URL parameter.

**- How to verify it**
```sh
$ docker run alpine # or create a container in any other way you prefer
$ curl -s --unix-socket /var/run/docker.sock 'http://localhost/system/df?types=image&type=container' | jq # should see image and container data
$ curl -s --unix-socket /var/run/docker.sock 'http://localhost/system/df?type=container&type=invalid' | jq # should see an error:
{
  "message": "unknown object type: invalid"
}
$ curl -s --unix-socket /var/run/docker.sock 'http://localhost/system/df?type=container' | jq # should see only container data
$ curl -s --unix-socket /var/run/docker.sock 'http://localhost/system/df' | jq # should see data, just like before
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

`/system/df` endpoint now supports `type` URL parameter, which gives control over object types to compute disk usage of

Commit:
API,daemon: support `type` URL parameter to /system/df

Let clients choose object types to compute disk usage of.

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>
Co-authored-by: Sebastiaan van Stijn <github@gone.nl>

rvolosatovs commented on daemon/disk_usage.go:
```diff
@@ -5,46 +5,60 @@ import (
 
 // SystemDiskUsage returns information about the daemon data disk usage
-func (daemon *Daemon) SystemDiskUsage(ctx context.Context) (*types.DiskUsage, error) {
+func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (*types.DiskUsage, error) {
```
I looked into possibility of using functional options here, but the issue is that it would require us to define the exact same struct of `bool`s in `api/server/router/system` in order to make things like `WithContainers()` work, since internally they would have to set a boolean field. Of course the internal struct could be abstracted, and getters could be exposed for the implementation to use, but overall, I do not see benefits of using this approach in this particular case, where a simple struct of bools suffices.

thaJeztah replied:
Yeah, I guess if we want those, one option _could be_ to make the signature for the option `func(*types.DiskUsage) error`, and the functions that are passed would do the actual collecting, e.g. (quick blurb)

```go
func WithContainerUsage() UsageOpt
	return func(usage *types.DiskUsage) error {
		// Retrieve container list
		containers, err = daemon.Containers(&types.ContainerListOptions{
			Size: true,
			All:  true,
		})
		if err != nil {
			return nil, fmt.Errorf("failed to retrieve container list: %v", err)
		}
		usage.Containers = containers
		return nil
	}
}
```

thaJeztah replied:
Yes, mostly "theoretical", I don't think there's much benefit (if any) over the current approach. If we _would_ have to define it, we can't do so in the daemon (as the API should not import the daemon), so then the `daemon.Containers` would have to be `<backend>.Containers`, and the backend passed in.

rvolosatovs commented on api/server/router/system/system_routes.go:
```diff
@@ -90,36 +91,77 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+		})
 		if err != nil {
-			return pkgerrors.Wrap(err, "error getting build cache usage")
+			return pkgerrors.Wrap(err, "error getting system disk usage")
```
Note: a wrapper was missing here, added

thaJeztah replied:
Perhaps (the "system" in this function's name / route is because it's a `docker system df` command, but perhaps "system disk" may cause some confusion).

```suggestion
			return pkgerrors.Wrap(err, "error getting disk usage")
```

That said, we should look if it might have been on purpose, and if the extra string adds sufficient value; there's some inconsistencies in `SystemDiskUsage` itself as well; some errors are not wrapped (but can return a "typed" error), other's are discarding underlying errors by creating a new error (without `Wrap` or `%w`); https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/disk_usage.go#L15
https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/disk_usage.go#L25
https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/disk_usage.go#L31

So this would look (for the above) like:

```
error getting system disk usage: a disk usage operation is already running
error getting system disk usage: failed to retrieve container list: <underlyihg error message>
error getting system disk usage: failed to retrieve image list: <underlyihg error message>
```


The "failed to retrieve container list" and "failed to retrieve image list" cases may actually be wrong, as (looking at the "images" case, the underlying errors may be of a specific `errdef` type; https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/daemon/images/images.go#L53-L61

https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/api/types/filters/parse.go#L250-L256

rvolosatovs replied:
I think it's better to dive into this in another PR, removing

thaJeztah replied:
Yes, agreed; should be looked into, but let's take the "safe route" for now

thaJeztah commented on api/server/router/image/backend.go:
```diff
@@ -22,7 +22,7 @@ type Backend interface {
 	ImageDelete(imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error)
 	ImageHistory(imageName string) ([]*image.HistoryResponseItem, error)
-	Images(imageFilters filters.Args, all bool, withExtraAttrs bool) ([]*types.ImageSummary, error)
+	Images(opts types.ImageListOptions, withContainerCount bool) ([]*types.ImageSummary, error)
```
Curious; was there a reason keeping `withContainerCount` separate, and not add it to the `types.ImageListOptions` ?

thaJeztah replied:
On a side-note; as we're changing the interface here, wondering if we should take the opportunity to also add a `ctx context.Context` now. (could be a placeholder for now)

rvolosatovs replied:
> Curious; was there a reason keeping `withContainerCount` separate, and not add it to the `types.ImageListOptions` ?

Since this parameter is not exposed via the HTTP API and only used internally, I felt that it is wrong to add it to `api/types`.
In the first iteration I did add it in the `types.ImageListOptions` though. Shall I bring it back in?

thaJeztah replied:
Yes, I think it makes the interface cleaner, but looks like I should've commented this on https://github.com/moby/moby/pull/42531 (which made the initial change 😅)

rvolosatovs replied:
No problem, added https://github.com/moby/moby/pull/42531/commits/dd61261e5354d97c39bf3be6224c3dcfe6441430

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -90,49 +101,80 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+			var err error
+			buildCache, err = s.builder.DiskUsage(ctx)
+			if err != nil {
+				return pkgerrors.Wrap(err, "error getting build cache usage")
```
If we _do_ want to add the wrapping above, this one should also be included / have the same prefix;

```suggestion
				return pkgerrors.Wrap(err, "error getting disk usage: error getting build cache usage")
```

I'm wondering why this code is separate here, and why it was not included in `backend.SystemDiskUsage`; perhaps because they didn't want the builder to be added to the daemon struct 🤔 It would make things somewhat cleaner here though.

rvolosatovs replied:
> I'm wondering why this code is separate here, and why it was not included in `backend.SystemDiskUsage`; perhaps because they didn't want the builder to be added to the daemon struct thinking It would make things somewhat cleaner here though.

I'd guess it has something to do with https://github.com/moby/moby/blob/5b708507ff45019673d0ac92797edcdcc2908f00/daemon/disk_usage.go#L15-L18

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -90,49 +101,80 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+		for _, s := range typesStrs {
+			for _, typ := range strings.Split(s, ",") {
+				switch typ {
+				case "container", "containers":
```
I'd omit the multiple variants; it's better to be strict here. It's an API, so it's ok to have a fixed list of options (if we want aliases for "convenience", those can be added in the CLI if we think the UX can be improved by adding them.

Perhaps would be nice to have `const`s for these values, so that a client can use those 🤔

rvolosatovs commented:
Blocking #42560

fredericdalleau commented on api/server/router/system/system_routes.go:
```diff
@@ -89,51 +100,89 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+
+const (
+	// ContainerObject represents a container DiskUsageObject.
+	ContainerObject DiskUsageObject = "container"
```
nit: one of the example you provided used `containers`, `images`, `volumes`. I personnaly prefer with the s

rvolosatovs replied:
https://github.com/moby/moby/pull/42559/commits/72be7d4c6ada92271eb8a6f5f562427efa815ab6

fredericdalleau reviewed (commented):
LGTM

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -89,51 +100,89 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+	// VolumeObject represents a volume DiskUsageObject.
+	VolumeObject DiskUsageObject = "volumes"
+	// BuildCacheObject represents a build-cache DiskUsageObject.
+	BuildCacheObject DiskUsageObject = "builds"
```
Perhaps `build-cache` instead of `build` for this one 🤔

rvolosatovs replied:
The reason I went for `build` originally is because we also have e.g. `/build/*` routes responsible for this object type, along with `/images/*`, `/volumes/*` etc.
IMO `build` is more consistent in this case, do you disagree?

thaJeztah replied:
Let me ping @tonistiigi if he has preference on this.

My main train of thought was; "can we expect other build-related objects at some point?" (idk, "background build jobs" or so) and in that case, would `build` (instead of `build-cache`) be the right term.


Naming is hard 😂

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -21,6 +22,16 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+type invalidRequestError struct {
```
Actually wondering why we added this one in the first place, instead of just using `errdefs.InvalidParameter()` utility (which should give the same).

Anyway, that's something to look at separately I guess. (Perhaps from that perspective, keep the first commit out of this PR)

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8343,6 +8343,16 @@ paths:
           schema:
             $ref: "#/definitions/ErrorResponse"
+      parameters:
+        - name: "types"
+          in: "query"
+          description: |
+            Object types, for which to compute and return data. For example, `containers,images`.
+          type: "array"
```
Remind me to check if this is the right description for a query-parameter that can be passed multiple times

rvolosatovs replied:
https://swagger.io/docs/specification/2-0/describing-parameters/#array

thaJeztah replied:
Ah, thanks! TIL about `collectionFormat`.

Looks like the current implementation (allowing both multiple params _and_ comma-separated values), cannot be described in Swagger. 

So it would require this to have either `collectionFormat: csv` (single, comma-separated values `?param=foo,bar,baz`):

```suggestion
        - name: "types"
          in: "query"
          description: |
            Object types, for which to compute and return data. For example, `containers,images`.
          type: "array"
          collectionFormat: csv
```


Or require `collectionFormat:multi` (multiple params, each with one option `?param=foo&param=bar&param=baz`):

```suggestion
        - name: "types"
          in: "query"
          description: |
            Object types, for which to compute and return data. For example, `containers,images`.
          type: "array"
          collectionFormat: multi
```

thaJeztah commented on api/server/router/system/backend.go:
```diff
@@ -10,12 +10,24 @@ import (
 
+// DiskUsageOptions holds parameters for system disk usage query.
+type DiskUsageOptions struct {
+	// Containers controls whether container disk usage should be computed.
+	Containers bool
+
+	// Images controls whether image disk usage should be computed.
+	Images bool
+
+	// Volumes controls whether volume disk usage should be computed.
+	Volumes bool
```
So, I was looking if an alternative could be  to use a `map[DiskUsageObject]bool` to indicate which objects to collect, then iterate over that list. 

I did some quick dabbling with that idea; let me push that to a branch, and we can chat Tomorrow

rvolosatovs commented on api/server/router/system/system_routes.go:
```diff
@@ -89,51 +100,89 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+
+const (
+	// ContainerObject represents a container DiskUsageObject.
+	ContainerObject DiskUsageObject = "containers"
```
TBH, I also prefer singular (that's also what was in the original implementation)
@fredericdalleau any comments on this? https://github.com/moby/moby/pull/42559#discussion_r660805286

rvolosatovs commented on api/server/router/system/system_routes.go:
```diff
@@ -89,51 +100,89 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
-	var builderSize int64
-	for _, b := range buildCache {
-		builderSize += b.Size
+	if getBuildCache {
```
Indeed, did this more to make it obvious it is not used, will change!

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8343,6 +8343,16 @@ paths:
+          description: |
+            Object types, for which to compute and return data. For example, `containers,images`.
+          type: "array"
+          uniqueItems: true
```
Looks like this is not "strictly" needed (as we are able to handle duplicates). Guess it doesn't  _hurt_, but not strictly "true" 😅

rvolosatovs commented:
@thaJeztah PTAL, I think I came up with a reasonable compromise between the original implementation and https://github.com/moby/moby/commit/eb499a67fa31b574bd38634ff9c86a3b69d93995

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -89,25 +89,60 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+	var objectTypes map[DiskUsageObject]struct{}
+	if ss, ok := r.Form["types"]; ok {
+		objectTypes = make(map[DiskUsageObject]struct{}, 4)
+		for _, s := range ss {
```
nit: the `s` variable collides with the `s *systemRouter` receiver, so perhaps we should pick a different name, e.g. 

```suggestion
		for _, ts := range ss {
```

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -89,25 +89,60 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 }
 
+// DiskUsageObject represents an object type used for disk usage query filtering.
+type DiskUsageObject string
+
+const (
```
I just realised; should we move these consts out of the `api/server` and into `api/types` to make them available to clients (in case they want to use them)?

Did a quick move with that (not a fan of that package being just `types` (but we're missing a `system` package in `api/types` currently), but at least it makes it consumable by clients.

Diff / patch below

<details>

```patch
diff --git a/api/server/router/system/backend.go b/api/server/router/system/backend.go
index 2ddd5e0018..2d18d389da 100644
--- a/api/server/router/system/backend.go
+++ b/api/server/router/system/backend.go
@@ -13,7 +13,7 @@ import (
 // DiskUsageOptions holds parameters for system disk usage query.
 type DiskUsageOptions struct {
 	// ObjectTypes to include when collecting usage.
-	ObjectTypes map[DiskUsageObject]struct{}
+	ObjectTypes map[types.DiskUsageObject]struct{}
 }
 
 // Backend is the methods that need to be implemented to provide
diff --git a/api/server/router/system/system_routes.go b/api/server/router/system/system_routes.go
index b67fa2d870..e451e1ab7c 100644
--- a/api/server/router/system/system_routes.go
+++ b/api/server/router/system/system_routes.go
@@ -89,31 +89,17 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 	return httputils.WriteJSON(w, http.StatusOK, info)
 }
 
-// DiskUsageObject represents an object type used for disk usage query filtering.
-type DiskUsageObject string
-
-const (
-	// ContainerObject represents a container DiskUsageObject.
-	ContainerObject DiskUsageObject = "container"
-	// ImageObject represents an image DiskUsageObject.
-	ImageObject DiskUsageObject = "image"
-	// VolumeObject represents a volume DiskUsageObject.
-	VolumeObject DiskUsageObject = "volume"
-	// BuildCacheObject represents a build-cache DiskUsageObject.
-	BuildCacheObject DiskUsageObject = "build-cache"
-)
-
 func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
 	if err := httputils.ParseForm(r); err != nil {
 		return err
 	}
 
-	var objectTypes map[DiskUsageObject]struct{}
+	var objectTypes map[types.DiskUsageObject]struct{}
 	if ss, ok := r.Form["types"]; ok {
-		objectTypes = make(map[DiskUsageObject]struct{}, 4)
-		for _, s := range ss {
-			switch typ := DiskUsageObject(s); typ {
-			case ContainerObject, ImageObject, VolumeObject, BuildCacheObject:
+		objectTypes = make(map[types.DiskUsageObject]struct{}, 4)
+		for _, ts := range ss {
+			switch typ := types.DiskUsageObject(ts); typ {
+			case types.ContainerObject, types.ImageObject, types.VolumeObject, types.BuildCacheObject:
 				objectTypes[typ] = struct{}{}
 			default:
 				return invalidRequestError{fmt.Errorf("unknown object type: %s", typ)}
@@ -133,7 +119,7 @@ func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter,
 	})
 
 	var buildCache []*types.BuildCache
-	if _, ok := objectTypes[BuildCacheObject]; ok || len(objectTypes) == 0 {
+	if _, ok := objectTypes[types.BuildCacheObject]; ok || len(objectTypes) == 0 {
 		eg.Go(func() error {
 			var err error
 			buildCache, err = s.builder.DiskUsage(ctx)
diff --git a/api/types/types.go b/api/types/types.go
index e3a159912e..2e2b02e62a 100644
--- a/api/types/types.go
+++ b/api/types/types.go
@@ -530,6 +530,20 @@ type ShimConfig struct {
 	Opts   interface{}
 }
 
+// DiskUsageObject represents an object type used for disk usage query filtering.
+type DiskUsageObject string
+
+const (
+	// ContainerObject represents a container DiskUsageObject.
+	ContainerObject DiskUsageObject = "container"
+	// ImageObject represents an image DiskUsageObject.
+	ImageObject DiskUsageObject = "image"
+	// VolumeObject represents a volume DiskUsageObject.
+	VolumeObject DiskUsageObject = "volume"
+	// BuildCacheObject represents a build-cache DiskUsageObject.
+	BuildCacheObject DiskUsageObject = "build-cache"
+)
+
 // DiskUsage contains response of Engine API:
 // GET "/system/df"
 type DiskUsage struct {
diff --git a/daemon/disk_usage.go b/daemon/disk_usage.go
index 9d345c905f..024fe7f3b5 100644
--- a/daemon/disk_usage.go
+++ b/daemon/disk_usage.go
@@ -23,20 +23,18 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsage
 		images     []*types.ImageSummary
 		layersSize int64
 		volumes    []*types.Volume
-
-		err error
+		err        error
 	)
-	objectTypes := opts.ObjectTypes
-	if len(objectTypes) == 0 {
-		objectTypes = map[system.DiskUsageObject]struct{}{
-			system.ContainerObject: {},
-			system.ImageObject:     {},
-			system.VolumeObject:    {},
+	if len(opts.ObjectTypes) == 0 {
+		opts.ObjectTypes = map[types.DiskUsageObject]struct{}{
+			types.ContainerObject: {},
+			types.ImageObject:     {},
+			types.VolumeObject:    {},
 		}
 	}
-	for typ := range objectTypes {
+	for typ := range opts.ObjectTypes {
 		switch typ {
-		case system.ContainerObject:
+		case types.ContainerObject:
 			// Retrieve container list
 			containers, err = daemon.Containers(&types.ContainerListOptions{
 				Size: true,
@@ -46,7 +44,7 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsage
 				return nil, fmt.Errorf("failed to retrieve container list: %v", err)
 			}
 
-		case system.ImageObject:
+		case types.ImageObject:
 			// Get all top images with extra attributes
 			images, err = daemon.imageService.Images(filters.NewArgs(), false, true)
 			if err != nil {
@@ -58,13 +56,13 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsage
 				return nil, err
 			}
 
-		case system.VolumeObject:
+		case types.VolumeObject:
 			volumes, err = daemon.volumes.LocalVolumesSize(ctx)
 			if err != nil {
 				return nil, err
 			}
 
-		case system.BuildCacheObject:
+		case types.BuildCacheObject:
 			// NOTE: build-cache data usage is currently collected via
 			// direct call to buildkit.Builder.DiskUsage in systemRouter.getDiskUsage().
 
```

</details>

thaJeztah commented on daemon/disk_usage.go:
```diff
@@ -5,46 +5,77 @@ import (
+		images     []*types.ImageSummary
+		layersSize int64
+		volumes    []*types.Volume
 
```
minor nit: stray empty line here; might as well have them all together

thaJeztah commented on daemon/disk_usage.go:
```diff
@@ -5,46 +5,77 @@ import (
-		return nil, fmt.Errorf("failed to retrieve image list: %v", err)
+		err error
+	)
+	objectTypes := opts.ObjectTypes
+	if len(objectTypes) == 0 {
+		objectTypes = map[system.DiskUsageObject]struct{}{
```
Nit/idea: I guess we no longer strictly need the intermediate variable, and could simplify to:

```suggestion
	if len(opts.ObjectTypes) == 0 {
		opts.ObjectTypes = map[system.DiskUsageObject]struct{}{
```

thaJeztah reviewed (commented):
Thanks! Yes, I like this version a lot!

Left one idea (moving the `DiskUsageObject` consts), and some nits (those are definitely no show-stoppers); posted a diff with those changes, in case you want to apply them.

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
@fredericdalleau @tianon @cpuguy83 PTAL

cpuguy83 commented on daemon/disk_usage.go:
```diff
@@ -5,46 +5,75 @@ import (
+			types.VolumeObject:    {},
+		}
 	}
+	for typ := range opts.ObjectTypes {
```
This loop adds some unnecessary indentation.
Instead can we do

```go
if opts.ObjectTypes[types.ContainerObject] {
    containers, err = ...
}
```

thaJeztah replied:
Won't that make it a very longe `if / else if / else if / else if / else` ? Wondering if that makes it more readable.

cpuguy83 replied:
Why? If we need containers then get them, otherwise there shouldn't be anything to do.

thaJeztah replied:
Ah, yes you're right. Didn't think straight; Was thinking "we don't return, so we must handle the rest in "else"

rvolosatovs replied:
If we don't iterate over map and we use it like a struct with booleans, we might as well just use a struct with booleans instead, which is simpler and less error-prone.
E.g. see original implementation: https://github.com/moby/moby/blob/310910eba3f1417fcddc091f98941afcefa23df4/daemon/disk_usage.go#L22-L57

cpuguy83 commented on api/server/router/system/system_routes.go:
```diff
@@ -90,24 +90,45 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
+		return err
+	}
+
+	var objectTypes map[types.DiskUsageObject]struct{}
```
Nit, it seems like you really want a `map[string]bool` here.

rvolosatovs replied:
Well, I want a set of enums, and the most idiomatic way to do that in Go is using a `struct{}`-valued map AFAIK.
Making this a map of `bool`s would introduce ambiguity - i.e. what does `containers`: `false` mean? If it should not be computed, it should just not be there.
Also, we get basically just an over-complicated `struct` with booleans, so, again, might as well just use a struct instead.

cpuguy83 commented on api/server/router/system/system_routes.go:
```diff
@@ -90,24 +90,45 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
-		}
-		return nil
-	})
+	if _, ok := objectTypes[types.BuildCacheObject]; ok || len(objectTypes) == 0 {
```
Just like in the daemon code for disk usage, maybe have a default value that adds all the object types so `len(objectTypes)` is never `0` (which wouldn't make much sense from an API perspective).

thaJeztah replied:
The idea here was that the backend continues to have its own default (if no options passed, everything is returned).

It's a bit unfortunate that `build cache` is the odd one out here, and not provided by the daemon/backend (in which case this special handling wouldn't be needed here.

Wondering it it would make sense to integrate buildkit there instead of here (not sure)

rvolosatovs commented:
Given the review from @cpuguy83 should be just revert back to original implementation with a struct and a few booleans? https://github.com/moby/moby/commit/310910eba3f1417fcddc091f98941afcefa23df4
In my opinion, using a map here creates more problems than it solves and raises complexity. We could always refactor this later if necessary (e.g. a new object type is added).

rvolosatovs commented:
Reverted back to using a struct: https://github.com/moby/moby/compare/b118789ae203c747d92469929f3bc8eee8841c9e..9f28f390da9abf71012bfa431e03f2dd526d973e

thaJeztah reviewed (approved):
LGTM

rvolosatovs commented:
Since #42531 was merged, a merge conflict appeared - rebased to fix that.

thaJeztah commented:
I guess the merge-conflict was in docs/api/version-history.md, or was it elsewhere?

thaJeztah reviewed (approved):
still LGTM

rvolosatovs commented:
> I guess the merge-conflict was in docs/api/version-history.md, or was it elsewhere?

Yes, but also in `daemon/disk_usage.go` due to image listing method signature change introduced by #42531

cpuguy83 reviewed (approved):
LGTM

cpuguy83 reviewed (dismissed):
We just need integration tests to verify the new functionality.

rvolosatovs commented:
Changes since last iteration:
- Added an integration test
- Updated the PR to account for recent changes regarding formatting (e.g. https://github.com/moby/moby/pull/42607). In the current implementation the fields that were not computed are `null`, or `[]` if they were, but were empty.
- Ensure that `daemon.SystemDiskUsage` is called only if image/volume/container data is actually requested

rvolosatovs commented:
cc @thaJeztah if you want to take (yet) another look. Sorry it's taking so many reviews!

thaJeztah commented on integration/system/disk_usage_test.go:
```diff
@@ -0,0 +1,181 @@
+
+	doc = makeDocFunc("after container.Run")
+	runTests(t,
+		TestCase{
```
These test-cases are all the same, with the exception of the container counter? If so, should se just make a

```
expected := foo
for set := range []string{"before", "after"} {
    for tests {
        t.Run(set + test.doc ())
    }
    run container
    update expected
}
```

thaJeztah commented on integration/system/disk_usage_test.go:
```diff
@@ -0,0 +1,181 @@
+		}
+	}
+
+	makeDocFunc := func(prefix string) func(string) string {
+		return func(suffix string) string {
+			return fmt.Sprintf("%s/%s", prefix, suffix)
+		}
+	}
+
```
I think this is a bit too much; just inline `t.Run(bla+"/"+tc.doc)`

rvolosatovs commented:
@thaJeztah @cpuguy83 PTAL

thaJeztah reviewed (dismissed):
LGTM

thaJeztah commented:
Arf.. I should've looked at Jenkins looks like the test is failing 😞

rvolosatovs commented:
I usually make a few iterations and request review when it's ready for (repeated) review.

rvolosatovs commented:
It seems that there's no way to use the testing approach I chose, since the store is modified in parallel to the test execution unless we use `-p=1` like in #42594. Is there a simple way to setup the daemon with a custom store for a particular test? Or perhaps some way to acquire a global lock for the duration of the test?

cpuguy83 commented:
You can setup a separate daemon for the test:

```go
d := daemon.New(t, "--iptables=false")
defer d.Cleanup(t)
d.Start(t) // or d.StartWithBusybox if needed.
defer d.Stop(t)
```

rvolosatovs commented:
> You can setup a separate daemon for the test:
> 
> ```go
> d := daemon.New(t, "--iptables=false")
> defer d.Cleanup(t)
> d.Start(t) // or d.StartWithBusybox if needed.
> defer d.Stop(t)
> ```

Perfect, I followed this approach!

rvolosatovs commented:
Looks like 2 test instances still failed :thinking: . This time it seems to be limited to Windows:
```
disk_usage_test.go:16: [d9303c6a11edc] failed to start daemon with arguments [--data-root D:\gopath\src\github.com\docker\docker\bundles\tmp\TestDiskUsage\d9303c6a11edc\root --exec-root C:\Windows\TEMP\dxr\d9303c6a11edc --pidfile D:\gopath\src\github.com\docker\docker\bundles\tmp\TestDiskUsage\d9303c6a11edc\docker.pid --userland-proxy=true --containerd-namespace d9303c6a11edc --containerd-plugins-namespace d9303c6a11edcp --containerd /var/run/docker/containerd/containerd.sock --host unix://C:\Windows\TEMP\docker-integration\d9303c6a11edc.sock --debug --storage-driver overlay2 --iptables=false] : protocol not available
```

That is on `d.Start(t, "--iptables=false")`. Perhaps `--iptables=false` is not supported on Windows and that's why it fails? Should we check the OS before adding the flag? Is there an alternative one for Windows?

rvolosatovs commented:
Removed `--iptables=false` flag, since many other tests also don't use it. Added `t.Parallel()` as well, since we now use a separate daemon.

rvolosatovs commented:
Looks like `iptables` was not the problem:
```
disk_usage_test.go:18: [d81ab14b1027e] failed to start daemon with arguments [--data-root D:\gopath\src\github.com\docker\docker\bundles\tmp\TestDiskUsage\d81ab14b1027e\root --exec-root C:\Windows\TEMP\dxr\d81ab14b1027e --pidfile D:\gopath\src\github.com\docker\docker\bundles\tmp\TestDiskUsage\d81ab14b1027e\docker.pid --userland-proxy=true --containerd-namespace d81ab14b1027e --containerd-plugins-namespace d81ab14b1027ep --containerd /var/run/docker/containerd/containerd.sock --host unix://C:\Windows\TEMP\docker-integration\d81ab14b1027e.sock --debug --storage-driver overlay2] : protocol not available
```

It looks like tests that do `d.Start` are actually skipped on Windows, so I followed the same approach.

rvolosatovs commented:
@cpuguy83 @thaJeztah PTAL

thaJeztah reviewed (approved):
LGTM
